Basic UI styling changed to SASS + initial theming support

kubawolanin opened the pull request:
Inspired by [this thread](https://community.openhab.org/t/basic-ui-dark-theme/33603) I started to play around the Basic UI front-end. Treat this PR as a proposal and a discussion point, rather than a full-blown feature :-)

So far I've managed to:
* clean up the stylesheets, 
* add some CSS custom properties (variables) in `_theming.scss` file 
* and move from LESS to SCSS.

The reason for the last bullet point was to potentially utilize Material Design Lite in [scss version only](https://github.com/tomatau/mdl-scss). That would give us some more flexibility wrt styling (e.g. extend css classes with internal mdl components or create your own).

CSS Custom Properties, however, are a great way to support theming in the runtime. 
I've [applied them in HABPanel](https://github.com/openhab/org.openhab.ui.habpanel/pull/187) as well and amount of possibilities they gave to the UI theming is huge 😎 

~~**I couldn't find a way to test my work** within a openhab2 dev environment. I hope I will find some help here. I've run `gulp` task to build the front-end. Then I cleanend and run the Eclipse workspace.
None of my changes appeared on the running UI though.
I didn't find any information regarding building Basic UI specifically in [the documentation](https://eclipse.org/smarthome/documentation/features/ui/basic/readme.html).~~

Edit: It all works now :-) forgot to enable the plugin in Run configuration 🙄 Thx @resetnow 

Cheers!

Signed-off-by: Kuba Wolanin <hi@kubawolanin.com> (github: kubawolanin)

resetnow commented:
>I couldn't find a way to test my work 

Assuming you have imported orc.eclipse.smarthome.ui.basic bundle and enabled it in the run configuration, it should work the way you described. Try changing any file in `web/` directory and see if you changes appear in browser.

resetnow commented on extensions/ui/org.eclipse.smarthome.ui.basic/snippets/main.html:
```diff
@@ -61,7 +61,7 @@
 		</script>
 	</head>
-	<body class="mdl-color--grey-100 mdl-color-text--grey-700" data-sitemap="%sitemap%" data-page-id="%id%" data-icon-type="%icon_type%">
+	<body class="mdl-color-text--grey-700" data-sitemap="%sitemap%" data-page-id="%id%" data-icon-type="%icon_type%">
```
These classes may need to be left as is for browsers without CSS variables support — BasicUI still works with IE 11 and not-so-recent Android versions. This will introduce a graceful fallback.

https://caniuse.com/#feat=css-variables

kubawolanin replied:
CSS Custom Properties will fallback gracefully as well if you pass a second argument to them, e.g
```css
background: var(--body-bg, #f5f5f5);
```

And this is what I did in my follow-up commit.

kubawolanin commented:
Thanks @resetnow for the quick review!
I've added a fallback values to the CSS custom properties, see [here](https://github.com/kubawolanin/smarthome/blob/7493737811d4b20328d7baff364e59060a962fd2/extensions/ui/org.eclipse.smarthome.ui.basic/web-src/_theming.scss#L57).

I've also added a few themes already 😺 
So when you go to http://<yourESH>/paperui/index.html#/configuration/services and click on Basic UI configuration, you'll be able to change `theme` property to one of the following values:

```
default
dark
blue-grey
deep-orange
green
```

And here's a little sneak peek of the themes:

![basic-ui-themes](https://user-images.githubusercontent.com/2270505/29996944-360679fa-9008-11e7-8a4d-68da68389c71.png)

At this point I think I can remove "WIP" flag from the title and we can proceed with the proper review :-)

Cheers!

resetnow commented:
>I've added a fallback values to the CSS custom properties, see here.

I don't think that's how fallback works. I have tested <a href="https://codepen.io/vladislav/pen/vJbwrv">this page</a> in a couple of browsers and it doesn't work for me, am I doing this right?

<img src="https://user-images.githubusercontent.com/2057191/29997199-ff56a688-9015-11e7-84df-65999fc7c957.png" width="50%" />
<img src="https://user-images.githubusercontent.com/2057191/29997200-ff56d4aa-9015-11e7-8852-1f79a7aa543e.png" width="50%"/>
<img src="https://user-images.githubusercontent.com/2057191/29997216-5aabdaee-9016-11e7-89a4-851a559b09fb.png" width="50%"/>

I think a better solution would be specifying the same property twice, only the second time using the variable syntax.

```css
color: black;
color: var(--color);
```

kubawolanin commented:
@resetnow well spotted! I've added better fallback now. Thanks

ThomDietrich commented:
@kubawolanin may I ask about the new home icon in the top left corner (what/why) and if that maybe means that you've touched https://github.com/eclipse/smarthome/issues/3859 (careful, the issue is a bit confusing)

resetnow commented:
@ThomDietrich Judging by the lack of arrows this is a result of openHAB bundles not being picked up by runtime.

kubawolanin commented:
Hey @ThomDietrich 👋 
No, I haven't touched it :-) My browser complained about MaterialIcons file being missing from the source so I've quickly added it temporarily to my dev environment to see if it helps.
As you can see, the group frame's arrows are not rendering too.
Edit: just found out that the issue is known -> https://github.com/eclipse/smarthome/issues/3523

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.basic/ESH-INF/config/config.xml:
```diff
@@ -5,6 +5,18 @@
         http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
 
 	<config-description uri="ui:basic">
+		<parameter name="theme" type="text" required="true">
```
This is now a pretty different approach than we have in the Paper UI.
With the Paper UI, we allow ESH solutions to add a custom theme on top by adding a fragment. It is NOT a feature exposed to the user, though. I think this makes sense in terms of ensuring a "corporate identity", i.e. allowing solutions to make sure that their UIs carry their logo and match there general CI.

While I can understand that it can be useful to allow users to choose a light or dark theme, I am not so convinced, if we really need orange/blue/green color options (resp. rather allow solutions to add additional themes besides light/dark through a fragment).

kubawolanin replied:
I'll leave just Default and Dark then. We can work on the fragment solution later on if that's preferred.

kubawolanin replied:
Done

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.basic/gulpfile.js:
```diff
@@ -3,19 +3,19 @@
 	var
 		gulp = require("gulp"),
-		less = require("gulp-less"),
+		sass = require("gulp-sass"),
```
This is a new tool that we didn't use before, correct? If so, we will require a CQ and approval by Eclipse.

resetnow replied:
We could also stick with LESS unless it doesn't support some of the required features.

kubawolanin replied:
I'd suggest to go with SASS because of the fact that MDL uses it natively by default. In the future we might extend/modify MDL stylesheets and possibly load them directly from `package.json` instead of hosting them as a static asset. 

See my comment from the first post:
> The reason for the last bullet point was to potentially utilize Material Design Lite in scss version only. That would give us some more flexibility wrt styling (e.g. extend css classes with internal mdl components or create your own).

I believe this approach will give us more flexibility overall and now is a good time to change it (since there's not much stylesheet to be maintained and conversion from LESS was pretty straightforward).

What's a process of submitting CQs? Is that something a PR creator should do?

Cheers! :-)

kubawolanin replied:
@resetnow @kaikreuzer 🏓

resetnow replied:
I think eventually having MDL building from sources is a good point.
 
>Is that something a PR creator should do?

No, as far as I understand, some people from Eclipse org can do it, including @kaikreuzer

kaikreuzer replied:
Correct, this has to be done by project committers.
Just to clarify: This is a build-time only dependency, i.e. we are not distributing anything from that project, right?

kubawolanin replied:
@kaikreuzer that's right.

kaikreuzer commented:
I have created CQ 14162 for gulp-sass approval.

kubawolanin commented:
@kaikreuzer is there any update on the IP check of gulp-sass?
Thanks!

kaikreuzer commented:
Not yet, waiting for it...

kaikreuzer commented:
CQ has just been approved!
